src/notify.rs

```diff
@@ -0,0 +1,178 @@
+    should_send: Arc<AtomicBool>,
+    sender: mpsc::Sender<()>,
+}
+
+impl NotifyHandle {
+    /// Create a new [`NotifyHandle`] with a link to the associated count.
+    pub(crate) fn new(counter: Arc<AtomicUsize>) -> (NotifyHandle, NotifySender) {
+        // Create a new "rendezvous channel". Note that we don't
+        // buffer any data in the channel, so memory won't grow if
+        // no-one is receiving any data.
+        let (sender, receiver) = mpsc::channel();
+        let should_send = Arc::new(AtomicBool::new(false));
+        (
+            NotifyHandle {
+                receiver,
+                should_send: Arc::clone(&should_send),
+                counter,
+            },
+            NotifySender {
+                sender,
+                should_send,
+            },
+        )
+    }
+
+    /// Block the current thread until the condition is true. This is
+    /// different than spin-looping since the current thread will use channels
+    /// internally to be notified when the counter changes.
+    pub fn wait_until_condition(
+        &self,
+        condition: impl Fn(usize) -> bool,
+    ) -> Result<(), NotifyError> {
+        self.wait_until_condition_inner(condition, |_| self.receiver.recv())
+            .map_err(|e| match e {
+                mpsc::RecvError => NotifyError::Disconnected,
+            })
+    }
+
+    /// [`NotifyHandle::wait_until_condition`] with a timeout.
+    pub fn wait_until_condition_timeout(
+        &self,
+        condition: impl Fn(usize) -> bool,
+        timeout: Duration,
+    ) -> Result<(), NotifyTimeoutError> {
+        self.wait_until_condition_inner(condition, |elapsed| {
+            let remaining_time = if let Some(remaining_time) = timeout.checked_sub(elapsed) {
+                remaining_time
+            } else {
+                return Err(mpsc::RecvTimeoutError::Timeout);
+            };
+
+            self.receiver.recv_timeout(remaining_time)
+        })
+        .map_err(|e| match e {
+            mpsc::RecvTimeoutError::Disconnected => NotifyTimeoutError::Disconnected,
+            mpsc::RecvTimeoutError::Timeout => NotifyTimeoutError::Timeout,
+        })
+    }
+
+    fn wait_until_condition_inner<E>(
+        &self,
+        condition: impl Fn(usize) -> bool,
+        recv_with_elapsed: impl Fn(Duration) -> Result<(), E>,
+    ) -> Result<(), E>
+    where
+        E: FromDisconnected,
+    {
+        let start = Instant::now();
+
+        // Drain all messages in the channel before turning sends on again.
+        while let Ok(()) = self.receiver.try_recv() {}
+        self.should_send.store(true, Ordering::SeqCst);
+
+        macro_rules! return_if_condition {
+            () => {
+                if condition(self.counter.load(Ordering::SeqCst)) {
+                    self.should_send.store(false, Ordering::SeqCst);
+                    return Ok(());
+                }
+            };
+        }
+
+        return_if_condition!();
```

Comment:
Yeah, the count is matching the condition and never changes, then we need to check before blocking on the `recv()` call.

Additionally, it's important that we set `should_store` to true before checking the condition (and we use a strong ordering like SeqCst) otherwise we could lose some change that doesn't notify the channel.